autograd.cpp

```diff
@@ -207,6 +211,57 @@ variable_list Conv::forward(variable_list input) {
   return variable_list({out});
 }
 
+void BatchNorm::initialize_parameters() {
+  if (affine_) {
+    weight = this->add(Var(DefaultTensor(at::kFloat).tensor(num_features_), true), "weight");
+    bias = this->add(Var(DefaultTensor(at::kFloat).tensor(num_features_), true), "bias");
+  }
+
+  if (stateful_) {
+    running_mean = Var(DefaultTensor(at::kFloat).zeros({num_features_}), false);
+    running_var = Var(DefaultTensor(at::kFloat).ones({num_features_}), false);
+  }
+}
+
+void BatchNorm::reset_parameters() {
+  if (affine_) {
+    weight.data().uniform_();
+    bias.data().zero_();
+  }
+
+  if (stateful_) {
+    running_mean.data().zero_();
+    running_var.data().fill_(1);
+  }
+}
+
+variable_list BatchNorm::forward(variable_list inputs) {
+  auto& input = inputs[0];
+  auto& running_mean = (stateful_ ? this->running_mean : inputs[1]);
+  auto& running_var = (stateful_ ? this->running_var : inputs[2]);
+
+  if (train_) {
+    int64_t channels = input.sizes()[0];
```

Comment:
This block is just `input.numel() / input.size(1)`. Also, why is the assert necessary? In the case of numel() == 0 tensors? Why not just directly check for that then?